k256/src/schnorr.rs

```diff
@@ -0,0 +1,318 @@
+//! Taproot Schnorr signatures as defined in [BIP340].
+//!
+//! [BIP340]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
+
+#![allow(non_snake_case)]
+
+use crate::{
+    AffinePoint, FieldBytes, FieldElement, NonZeroScalar, ProjectivePoint, PublicKey, Scalar,
+};
+use elliptic_curve::{
+    bigint::U256,
+    ops::{LinearCombination, Reduce},
+};
+use sha2::{Digest, Sha256};
+use signature::{Error, Result};
+
+const AUX_TAG: &[u8] = b"BIP0340/aux";
+const NONCE_TAG: &[u8] = b"BIP0340/nonce";
+const CHALLENGE_TAG: &[u8] = b"BIP0340/challenge";
+
+/// Taproot Schnoor signature as defined in [BIP340].
+///
+/// [BIP340]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub struct Signature {
+    bytes: [u8; 64],
+}
+
+impl Signature {
+    /// Borrow the serialized signature as bytes.
+    pub fn as_bytes(&self) -> &[u8; 64] {
+        &self.bytes
+    }
+
+    /// Get the `r` component of this signature.
+    fn r(&self) -> FieldElement {
+        FieldElement::from_bytes(FieldBytes::from_slice(&self.bytes[..32])).unwrap()
+    }
+
+    /// Get the `s` component of this signature.
+    fn s(&self) -> NonZeroScalar {
+        NonZeroScalar::from_repr(*FieldBytes::from_slice(&self.bytes[32..])).unwrap()
+    }
+
+    /// Split this signature into its `r` and `s` components.
+    fn split(&self) -> (FieldElement, NonZeroScalar) {
+        (self.r(), self.s())
+    }
+}
+
+impl AsRef<[u8]> for Signature {
+    fn as_ref(&self) -> &[u8] {
+        self.as_bytes()
+    }
+}
+
+/// Taproot Schnorr signing key.
+#[derive(Clone)]
+pub struct SigningKey {
+    /// Secret key material
+    secret_key: NonZeroScalar,
+
+    /// Verifying key
+    verifying_key: VerifyingKey,
+}
+
+impl SigningKey {
+    /// Parse signing key from big endian-encoded bytes.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self> {
+        let secret_key = NonZeroScalar::try_from(bytes).map_err(|_| Error::new())?;
+        let public_key =
+            PublicKey::from_affine((ProjectivePoint::generator() * *secret_key).to_affine())
+                .map_err(|_| Error::new())?;
+
+        Ok(Self {
+            secret_key,
+            verifying_key: VerifyingKey { inner: public_key },
```

Comment:
I forgot to add a TODO in the code, but this should be constructed using `DecompactPoint::decompact` once it properly implements the `lift_x` function.